### PR TITLE
Add log-probability correction to the exchange rule.

### PR DESCRIPTION
**Pull request by Z-Denis**
This minimal PR prevents the sampler from overestimating the acceptance when using the exchange rule. This is especially relevant when samples are close to uniform. There, the rule acts trivially most of the time, yielding a high "displayed" acceptance although the chain's state did not update.

### CODE REVIEW

**Comment by gcarleo** [comment]
Ah ok but then I don't like the solution proposed, I'd rather implement this in the acceptance!

**Comment by gcarleo** [comment]
Just put an if in the acceptance, if the confirmations are identical you claim it as rejected. A much better solution would be to propose to exchange only configurations that are different

**Comment by gcarleo** [comment]
If you propose to exchange only different local states then you need to be careful with the actual Hastings correction, but it's a much cleaner solution in my view

**Comment by PhilipVinc** [comment]
> When achieving finite acceptance is critical, one can alternatively mask out clusters involving matching local states.

What do you mean? Right now this would only affect what is logged/displayed but not the chain.


> Just put an if in the acceptance, if the configurations are identical you claim it as rejected

Is that technically correct and are we fine with the change (@gcarleo, @wdphy16 )? 
moves that do nothing are rejected even if technically the issue lies in the proposal rule not proposing anything useful, not in the wavefunction itself.

>  A much better solution would be to propose to exchange only configurations that are different

I guess that's something different than what Zakari is proposing here, but I agree that would be useful (if possible to implement easily)

**Comment by gcarleo** [comment]
I think it is wrong to claim that equal configurations are rejected, so I am against hacking the sampler.... I would instead add a flag that allows one to restrict the proposal such that only different local states are exchanged

**Comment by Z-Denis** [comment]
> If you propose to exchange only different local states then you need to be careful with the actual Hastings correction, but it's a much cleaner solution in my view 

This is what I was proposing in a previous comment. It makes the acceptance slightly higher. I have some code for that, it's just slightly more expensive.

**Comment by PhilipVinc** [comment]
> This is what I was proposing in a previous comment. It makes the acceptance slightly higher. I have some code for that, it's just slightly more expensive.

I think it would be a good addition. Maybe hidden behind a flag (default true) such that it ca be disabled if neeeded,

**Comment by Z-Denis** [comment]
> What you need for the move to be efficient is to 1. Identify all sites that have different occupation numbers, say there are K of those 2. Pick a random k in K 3. For the proposed move compute K", 4. Use K and K" in the Hastings correction 

What I was doing locally is to keep all this precomputed-clusters logic and mask out irrelevant ones with a modified `jax.random.choice` function, finally send the logp to minus infty whenever no cluster with different local configurations is available.

**Comment by gcarleo** [comment]
The Hastings correction would be log(K) indeed, so if K is zero you don't accept, that's a clean solution

**Comment by Z-Denis** [comment]
This version implements the above logic although it is more expensive and when used within a `MultipleRule` yields only slightly higher acceptance:
```python
def transition(rule, sampler, machine, parameters, state, key, σ):
    n_chains = σ.shape[0]

    # pick a random cluster
    mask = σ[..., rule.clusters[:, 0]] != σ[..., rule.clusters[:, 1]]
    cluster_id = batch_choice(
        key, shape=(n_chains,), a=np.arange(rule.clusters.shape[0]), p=mask
    )
    n_clusters_σ = jnp.sum(mask, axis=-1)

    def scalar_update_fun(σ, cluster):
        # sites to be exchanged,
        si = rule.clusters[cluster, 0]
        sj = rule.clusters[cluster, 1]

        σp = σ.at[si].set(σ[sj])
        return σp.at[sj].set(σ[si])

    σp = jax.vmap(scalar_update_fun, in_axes=(0, 0), out_axes=0)(σ, cluster_id)
    mask = σp[..., rule.clusters[:, 0]] != σp[..., rule.clusters[:, 1]]
    n_clusters_σp = jnp.sum(mask, axis=-1)

    log_corr = jnp.where((n_clusters_σ == 0) & (n_clusters_σp == 0), -jnp.inf,
                         -jnp.log(n_clusters_σp) + jnp.log(n_clusters_σ))

    return σp, log_corr
```

**Comment by wdphy16** [comment]
I'm fine with either definition of the acceptance rate, and this is why the acceptance rate is not a metric as good as the autocorrelation time

**Comment by gcarleo** [comment]
thanks Zakari I think we should definitely include the last code you wrote as a comment (maybe do a separate PR or within this one?) enabling the option to have distinct configurations in cluster through a flag! (I would put the flag=False, to avoid a breaking change @PhilipVinc ?) 

Maybe @Z-Denis can you just add a comment to why you do rule.clusters[:, 0]] != σ[..., rule.clusters[:, 1] ? I am not sure I understand why there is no sigma on both sides of the !=

**Comment by PhilipVinc** [comment]
@gcarleo yes, fine by me hiding it behind a flag. 

So the final state of this would be (@gcarleo do you agree?): we keep the current logic .  We add an option to sample only from different configurations in exchange.

And ideally @Z-Denis makes another PR to modify `multiple rules` to pick rules that propose different states.

Correct?

**Comment by Z-Denis** [comment]
> Maybe @Z-Denis can you just add a comment to why you do rule.clusters[:, 0]] != σ[..., rule.clusters[:, 1] ? I am not sure I understand why there is no sigma on both sides of the !=

I am not sure I see what you mean, I indeed compare sigmas on both sides of the `!=` operator: `σ[..., rule.clusters[:, 0]] != σ[..., rule.clusters[:, 1]]`.

I will update the PR. In the code I tested, I am using a batched choice function which replicates the behaviour of `jax.random.choice`. I can add it somewhere in `netket.jax` or instead use the simplified version in `netket.sampler.autoreg` needed for the autoregressive sampling logic. Any preference?

**Comment by gcarleo** [comment]
> I am not sure I see what you mean, I indeed compare sigmas on both sides of the != operator: σ[..., rule.clusters[:, 0]] != σ[..., rule.clusters[:, 1]].

yeah sorry, I misread the code!

**Comment by gcarleo** [comment]
also, it's likely this is more expensive than the simpler exchange rule, but I would expect this is certainly not the bottleneck, and even on small systems (i.e. the canonical 20 spins Heisenberg case) there should be essentially no difference at all with the standard Exchange rule? however there will be a slight advantage since configurations will be less correlated

**Comment by PhilipVinc** [comment]
> I will update the PR. In the code I tested, I am using a batched choice function which replicates the behaviour of jax.random.choice. 

Yes please
